### DISCUSSION TITLE
print on closed filehandle error

### DISCUSSION THREAD

**Discussion by laurel-nave-powers** [sr320/course-fish546-2021#113] [Q&A]
Carrying on from a previous q and a: https://github.com/sr320/course-fish546-2021/discussions/98 
My script was running for hours and was continuously spitting out this error: 
<img width="855" alt="Screen Shot 2021-03-02 at 11 30 23 AM" src="https://user-images.githubusercontent.com/76967600/109703838-cb16d680-7b4a-11eb-907c-de8fca062878.png">
This goes on forever, and GitHub desktop won't even let me push the notebook through to my repo which is why I took a screenshot. 
I am in the process of googling what this may mean, but wanted to post in case someone has a quick answer.

**Comment by kubu4** [comment]
Please paste the full command when you have a sec.

**Reply by kubu4** [reply]
With that said, the error message is saying that it is unable to open the intended file.

Simplest explanation is usually that the target file does not have the proper permissions.

The other "simple" explanation is that the target file is already open/being used by another program. When that's the case, the computer "locks" that file to prevent other programs from modifying it while it's currently in use.

Would either of these be possible situations?

**Reply by laurel-nave-powers** [reply]
You can see the whole command here: https://github.com/fish546-2021/Laurel-genes/blob/main/code/03_rmdup.ipynb 
(note that the error under the command in this notebook is not up to date and has been resolved). 

I am not sure... how would I find out if it has the proper permissions? 
I don't think it is being used by another program but am not 100% sure.

**Reply by kubu4** [reply]
Use the following commands:

`ls -l ~/Desktop/GitHub/Laurel-genes/analyses/trimmed`

`ls -l ~/Desktop/GitHub/Laurel-genes/analyses/samplelist.txt`

The left column of the output will show you the read (`r`), write (`w`) and/or execute (`x`) permissions for the files. Maybe post a screenshot, if you don't mind.

Also, can you please post a link to the `rmdup.pl` file? I'll glance at the code in there to double-check things.

**Reply by laurel-nave-powers** [reply]
Ok, I will try that, thank you. 
Here is the script: https://github.com/fish546-2021/Laurel-genes/blob/main/scripts/rmdup.pl

**Reply by laurel-nave-powers** [reply]
Here is the output for the permissions commands: 
<img width="566" alt="Screen Shot 2021-03-02 at 12 15 19 PM" src="https://user-images.githubusercontent.com/76967600/109709044-f69cbf80-7b50-11eb-9796-e2749bd6ba32.png">
It looks like I have permission for reading and writing but not executing?

**Reply by kubu4** [reply]
> It looks like I have permission for reading and writing but not executing?

Yep, everything looks good there...

Alrighty, looking at the code, I think what you're seeing is normal and don't think it's an error. I think. The code is this:

```perl
my $preline = substr $line, 0, $prefixlen; # get prefix
	if ($preline ne $prefix) { # if prefix of current line is not the same as reads file that is being written into
		close $PREFIX if ($prefix ne "none"); # close filehandle of reads file that is being written into
		$prefix = $preline; # update the prefix
		open $PREFIX, ">>", "$merged_reads/$sample.$preline.fq"; # open filehandle of file with the updated prefix
	}
	print $PREFIX $line; # write reads to the filehandle opening reads file
}
close SORT;
close $PREFIX;
```

This is indicating that the filehandler _should_ be closed if the prefix doesn't match the preline.

Now, with that sort of figured out, you'll want to redirect all of that output to a separate file. Otherwise, all of that output will render that notebook entirely unusable, as the notebook file size will exceed your computer's memory size. You won't even be able to open it any more. Additionally, it'll become too large to upload to GitHub, which isn't horrible, but is annoying.

So, do you remember from the readings how to redirect `stderr` to a file? See if you can figure it out and if you have questions, just post them.

Also, what you're trying to do will likely take a long time; just the nature of processing high-throughput sequencing data...

**Reply by laurel-nave-powers** [reply]
Ok, so I read in Chapter 3 that if I don't need the output message I can redirect it to a log file like stderr.txt, or if I want it to disappear I can write it to /dev/null. The /dev/null option seems like the one i would want so that it's not slowed down with writing to a physical disk? 
With that being said, I'm a little confused with the book's actual code examples for this (see pages 43-44). Would I do something like this? 
`ls -l error.txt 2> error.stderr`
Also, would this be tagged at the end of my command calling the script? 
Thank you for your help!

**Reply by kubu4** [reply]
Yep. I'd capture the error stream instead of sending it to `/dev/null`. So, add `2> error.stderr` to the end of your command.

**Reply by laurel-nave-powers** [reply]
Ahhh ok thank you! I wasn't sure if the `ls -l` needed to be included or if that was just part of the book example. I will add this to the end of my command and let it run. Thanks!

**Reply by laurel-nave-powers** [reply]
Well it finished suspiciously fast and though there was no error it did not actually give me any files with duplicates removed. Not sure what I am doing wrong or where to look since it didn't give me errors. 
It let me upload the notebook this time at least: https://github.com/fish546-2021/Laurel-genes/blob/main/code/03_rmdup.ipynb

**Reply by kubu4** [reply]
That `ls -l` command you ran in your Jupyter Notebook should've spit out _something_, even if there isn't anything in there.

When a directory is empty, the output should be:

`total 0`

Not sure what's going on there.

**Reply by laurel-nave-powers** [reply]
That's what I thought too! I'm wondering if it is just taking a long time to appear? Maybe my computer is struggling.

**Reply by kubu4** [reply]
Try finding the output files.

Navigate to:

`/Users/laurelnave-powers/Desktop/GitHub/Laurel-genes`

Then run:

`find . -type f -name "*.rmdup.fq"`

Explanation of what this is doing:

`find`: program called find. Searches for stuff.

`.`: The directory that you want `find` to start looking in. A `.` represents the current directory.

`-type f`: Tells `find` to only look for files (`f`). Will ignore names in directories (`d`).

`-name "*.rmdup.fq"`: Tells find that the name of the files will have the pattern inside the quotes. In this case, it will search for any filenames that end with `.rmdup.fq`

This will search recursively; meaning it will search the current directory for the filename pattern specified and then continue searching through any subdirectories until there are no more subdirectories/files.

**Reply by laurel-nave-powers** [reply]
I think there is something wrong with my jupyter notebook now because that didn't give me output either. I just exited out of jupyter lab and restarted it and it still didn't give me output. Not even a "no files with this name".

**Reply by laurel-nave-powers** [reply]
I put it in my normal terminal as well and it didn't give me anything. 
<img width="566" alt="Screen Shot 2021-03-03 at 10 15 30 AM" src="https://user-images.githubusercontent.com/76967600/109852203-63739080-7c09-11eb-8b6e-fff3b44ea2fa.png">

**Reply by kubu4** [reply]
`find` won't spit anything out if it doesn't find anything, so that's normal.

For kicks, can you run the `tail` command on your stderr log file and paste the results (or screencap)?

**Reply by laurel-nave-powers** [reply]
This is what it tells me when I try to run tail on error.stderr 
"Can't open perl script "rmdup.pl": No such file or directory". Not sure why or how to include the whole perl script into just a tail command on a file?

**Reply by kubu4** [reply]
Please post command used and/or screencap.

**Reply by laurel-nave-powers** [reply]
<img width="569" alt="Screen Shot 2021-03-03 at 10 25 47 AM" src="https://user-images.githubusercontent.com/76967600/109853527-d3cee180-7c0a-11eb-811c-0e6b438a79c2.png">

**Reply by kubu4** [reply]
OK, that's actually the entire contents of your `error.stderr` file. So, seems like the command actually didn't work, despite printing all of that stuff to the Jupyter Notebook...

**Reply by kubu4** [reply]
Try complete shutdown of Jupyter Notebook (including the Terminal where it's running) and try again. Let's see what happens!

**Reply by laurel-nave-powers** [reply]
Alright, I did a complete shutdown and got some different output. The only thing that changed is the shutdown and a change in the location of the error.stderr. Still no actual files as output. The new error.stderr has different output now as well when I run tail. I will look at the lines indicated to see if that gives some clues as to what is happening. 
See this updated notebook for details: https://github.com/fish546-2021/Laurel-genes/blob/main/code/03_rmdup.ipynb
